### PR TITLE
Fix broken nested imports.

### PR DESCRIPTION
**Pull request by drewloomer**

### CODE REVIEW

**Comment by yourfavorite** [comment]
I have tested this and it seems to fix the issue with Symbols that have `/`s in them getting lost.

**Comment by yourfavorite** [comment]
I've been testing this some more and it seems to work inconsistently. It worked well on some random test files and then I tested on a production file. It worked at first but has now stopped working and I can't seem to figure out the cause. I'm currently digging in and trying to add log statements to see if I can't contribute to fixing the issue, but as I don't know CocoaScript I am a bit lost haha.

**Comment by yourfavorite** [comment]
Here are some logs. I have added 2 log statements prefaced with `ISD:`. One is inside of the `addSymbol()` function. It just logs the `symbolName` and `symbolID`. Then I added a log statement inside of the if statement check for MSSymbolInstance inside of `getNestedSymbolMasters()` that logs `layers.objectAtIndex(i)`. I don't really know what exactly any of this means but I figure I should see matching IDs in these? (Sorry if that's horribly incorrect). At any rate, if there are other places I can add log statements that would provide valuable information, please let me know.

For the import, my imported symbol names are:
```
! Logo
! Nav / ! Nav
! Nav / Home
! Nav / More
! Nav / Schedule
! Nav / VODs
! Nav / Watch
```

! Nav / ! Nav is a symbol that includes all of the other symbols inside of it.
! Logo does not bug out, but all of the other symbols which are icons for the nav bar bug out.

```
default	15:27:38.668768 -0700	Sketch	ISD: Adding ! Logo ||| AF155EC4-52FB-41A3-8C1E-B1FD834836C1
default	15:27:38.688108 -0700	Sketch	ISD: Adding ! Nav / ! Nav ||| 8762A2E8-187C-440D-B086-867D9486A8CD
default	15:27:38.689042 -0700	Sketch	ISD: Instance of <MSSymbolInstance: 0x7ffb9666edc0> Logo (94A1D376-70D6-43AB-83E5-D8950340AF2F)
default	15:27:38.689674 -0700	Sketch	ISD: Adding ! Logo ||| AF155EC4-52FB-41A3-8C1E-B1FD834836C1
default	15:27:38.726484 -0700	Sketch	ISD: Adding ! Nav / Home ||| ECDDD8FF-D06B-4195-AD9F-67994C898ADB
default	15:27:38.743183 -0700	Sketch	ISD: Adding ! Nav / More ||| 9A803D57-CE42-4A38-8541-21BC8771E033
default	15:27:38.758902 -0700	Sketch	ISD: Adding ! Nav / Schedule ||| 5F0AC9F7-4ED6-483B-90B3-464FF092FE3F
default	15:27:38.777500 -0700	Sketch	ISD: Adding ! Nav / VODs ||| 0544DAAD-565F-40B8-B240-A58FC01F9808
default	15:27:38.793589 -0700	Sketch	ISD: Adding ! Nav / Watch ||| CB387EC8-7AD3-43AD-91BF-0A09ED8D2381
default	15:27:38.810515 -0700	Sketch	IS: <MSSymbolMaster: 0x7ffb96953e20> ! Logo (C6314D4F-6668-4AEF-B686-B4A230391E7C)
default	15:27:38.810784 -0700	Sketch	IS: undefined
default	15:27:38.817020 -0700	Sketch	IS: <MSSymbolMaster: 0x7ffb966517c0> ! Nav / ! Nav (65493819-A286-4623-8F1E-8B1CA388B60D)
default	15:27:38.817255 -0700	Sketch	IS: <MSSymbolMaster: 0x7ffb96953e20> ! Logo (C6314D4F-6668-4AEF-B686-B4A230391E7C)
default	15:27:38.818306 -0700	Sketch	IS: <MSSymbolMaster: 0x7ffb966d8370> ! Logo (A14AE493-40EA-47A4-B5DD-F372E18460FA)
default	15:27:38.818476 -0700	Sketch	IS: <MSSymbolMaster: 0x7ffb966517c0> ! Nav / ! Nav (65493819-A286-4623-8F1E-8B1CA388B60D)
default	15:27:38.819704 -0700	Sketch	IS: <MSSymbolMaster: 0x7ffb921e0ae0> ! Nav / Home (290E16D3-728D-4A53-BB04-7967E3CC3518)
default	15:27:38.819887 -0700	Sketch	IS: <MSSymbolMaster: 0x7ffb966d8370> ! Logo (A14AE493-40EA-47A4-B5DD-F372E18460FA)
default	15:27:38.820582 -0700	Sketch	IS: <MSSymbolMaster: 0x7ffb96b65920> ! Nav / More (211D950E-4178-4795-A00D-F868A187442C)
default	15:27:38.820763 -0700	Sketch	IS: <MSSymbolMaster: 0x7ffb921e0ae0> ! Nav / Home (290E16D3-728D-4A53-BB04-7967E3CC3518)
default	15:27:38.821463 -0700	Sketch	IS: <MSSymbolMaster: 0x7ffb968a1620> ! Nav / Schedule (BB4B518F-7C08-40A6-B72E-A471CB0105DA)
default	15:27:38.821669 -0700	Sketch	IS: <MSSymbolMaster: 0x7ffb96b65920> ! Nav / More (211D950E-4178-4795-A00D-F868A187442C)
default	15:27:38.822580 -0700	Sketch	IS: <MSSymbolMaster: 0x7ffb96937b60> ! Nav / VODs (A2E69D40-47C1-426F-A00E-FEC57ED0EDA0)
default	15:27:38.822811 -0700	Sketch	IS: <MSSymbolMaster: 0x7ffb968a1620> ! Nav / Schedule (BB4B518F-7C08-40A6-B72E-A471CB0105DA)
default	15:27:38.823614 -0700	Sketch	IS: <MSSymbolMaster: 0x7ffb922ed0e0> ! Nav / Watch (7B4FBB12-F2D1-48A0-B071-8EEE9AE51BA5)
default	15:27:38.823799 -0700	Sketch	IS: <MSSymbolMaster: 0x7ffb96937b60> ! Nav / VODs (A2E69D40-47C1-426F-A00E-FEC57ED0EDA0)
```

**Comment by yourfavorite** [comment]
Another minor bug I've noticed has to do with how positioning of symbols on the symbols page happens. All of my symbols in both files are organized Alphabetically so `! Logo` appears first and that symbol is referenced inside of `! Nav / ! Nav` (had to hide the logo for NDA but imagine it is positioned just inside the top of the tall black Nav Bar). You'll notice on the left that the symbol name is `! Logo` and on the right `! xLogo` which causes it to come alphabetically after the `! Nav / ! Nav` symbol. It appears that somewhere in the loops it positions the symbol each time it is referenced until the actual symbol gets positioned. Hence the gap in the image on the left. Logo gets positioned right after the Nav symbol because it is referenced inside of Nav. Then it gets positioned again just below where it was previously. Alright, I think that's the best I can explain this in text. Please let me know if this makes sense or if you have any questions.


import-symbol

**Comment by yourfavorite** [comment]
Did a little more research and I have figured out that the issue is with the file I'm trying to import symbols from. I don't know why it isn't working (or what I did to cause it), but I've scrubbed the file of sensitive data and thrown in some crap icons. Dropbox link below. Feel free to play with it.

Steps to reproduce:
1) Create a new sketch document.
2) Import by Name > Global UI.sketch

https://www.dropbox.com/s/oxtde6s1tetz6qf/Global%20UI.sketch?dl=0

**Comment by yourfavorite** [comment]
Some additional info that might help is the latest plugin from Lingo which does symbol importing very well. From my first look it appears as though the plugin is not obfuscated so its possible some inspiration could be pulled from there. https://www.lingoapp.com/sketch/

**Comment by yourfavorite** [comment]
So I spent some time learning cocoascript (as much as one can considering how poorly all of sketch plugin creation is documented)... And I think I've figured out the problem.

`getNestedSymbolInstances()` loops through the root of the symbol looking for instances of MSSymbolInstance. This works great if all nested symbols are at the root level. However if there are symbols inside of a Group/Folder, they don't get linked because `getNestedSymbolInstances()` sees a className of MSLayerGroup and ignores everything inside of the Group. So if that function is updated to loop recursively into Groups, this issue should be fixed.

**Comment by yourfavorite** [comment]
So I think I've made an update that fixes the issue but I'm not sure if it is the best way to handle it. I have changed `getNestedSymbolInstances()` to the following:

```
function getNestedSymbolInstances(symbol, symbols) {
	var arr = [[NSMutableArray new] init];
	var layers = symbol.layers();
	var i = 0;
	var len = layers.count();
	for (; i < len; i++) {
		if (layers.objectAtIndex(i).className().isEqualToString('MSSymbolInstance')) {
			arr.push(layers.objectAtIndex(i));
		}

		if (layers.objectAtIndex(i).className().isEqualToString('MSLayerGroup')) {
			var groupArr = getNestedSymbolInstances(layers.objectAtIndex(i), symbols);
			arr = arr.concat(groupArr);
		}
	}
	return arr;
}
```

**Comment by drewloomer** [comment]
Nice work, @yourfavorite! I've incorporated your changes into this PR. I haven't had a chance to test it (and I simplified the code ever so slightly), but it looks good to me. If you can verify that it's working as expected, perhaps we can put this nagging issue to bed...

**Comment by yourfavorite** [comment]
@drewloomer Looks like there are two more issues that need to be resolved. And in my testing if we fix these two things, we're good to go.

1) It looks like you can't concat NSMutableArray. At least not with the js concat. Instead of merging the arrays, it creates an array of arrays. I'm guessing there's an Obj-C way of doing this but I couldn't figure it out. So for now I've written the following function to merge the two arrays. (it could prob be cleaned up)

```
function pushConcat(arrA, arrB) {
	arrB.forEach(function (b) {
		arrA.push(b);
	});

	return arrA;
}
```

Usage:
```
arr = pushConcat(arr, getNestedSymbolInstances(layers.objectAtIndex(i), symbols));
```

2) The recursive piece we added to getNestedSymbolInstances, just needs to be replicated for getNestedSymbolMasters.

**Comment by yourfavorite** [comment]
One last update is that turns out the proper way to concat NSMutableArray is to use `arr.addObjectsFromArray()` so the update would be:

`arr = arr.addObjectsFromArray(getNestedSymbolMasters(layers.objectAtIndex(i), symbols));`

**Comment by drewloomer** [comment]
Okay, @yourfavorite I've done a lot of thinking about this problem. The code in this PR is really getting nasty and I think I'm over-solving in a bunch of places. I've created a new branch that is a 99% rewrite but seems to fix the nested import issue in a variety of cases. I'm hoping you can test it out for me and report back: https://github.com/drewloomer/sketch-import-symbols/tree/nested-rewrite

**Comment by yourfavorite** [comment]
@drewloomer Just gave it a run and I'm getting an error:
```TypeError: null is not an object (evaluating 'getLayerValue(symbols.objectAtIndex(i), 'import_' + replaceBy).isEqualToString')
line: 198
sourceURL: /Users/cbritt/Library/Application Support/com.bohemiancoding.sketch3/Plugins/Sketch Import Symbols/ImportSymbols.sketchplugin/Contents/Sketch/importSymbols.js
column: 71
stack: getSymbol@/Users/cbritt/Library/Application Support/com.bohemiancoding.sketch3/Plugins/Sketch Import Symbols/ImportSymbols.sketchplugin/Contents/Sketch/importSymbols.js:198:71
/Users/cbritt/Library/Application Support/com.bohemiancoding.sketch3/Plugins/Sketch Import Symbols/ImportSymbols.sketchplugin/Contents/Sketch/importSymbols.js:108:35
forEach@[native code]........```

**Comment by drewloomer** [comment]
:sob: :gun: Which version of Sketch are you using?

**Comment by yourfavorite** [comment]
Sketch 46.1. I guess it is possible that because I have been using these files with the old version of the plugin that there are conflicts? Haven't dug into your code to see what you've changed in this rewrite.

Just dragged the symbols over to a new file and still get the same error. Not sure if layer params get moved over to new files via drag and drop or not.

**Comment by drewloomer** [comment]
Hrm, that shouldn't be an issue. The plugin is really just the manifest and js file.

I've tried this rewritten version of the plugin with the Dropbox file you provided above and it worked perfectly as far as I can see. Could you try it with that simplified file and see if you get the same error? 

Also, was that the full stack trace that you posted last time? It cut off before it could be useful in debugging the issue. If there is any way to get a full call list that would be hugely helpful.

I'm sorry to ask you to be the guinea pig on this. Thanks!

**Comment by yourfavorite** [comment]
@drewloomer Here are the two files I'm testing with. Import `Test master` into slave and it fails.

https://www.dropbox.com/s/zkeerjmn8cyrvg8/sketch-import-test.zip?dl=0

**Comment by yourfavorite** [comment]
One other note while you're at it. After making updates to your previous pull request to fix the nested import issues, I made a tweak to the plugin to make it more flexible. I'm happy to take a crack at it again once you finish your rewrite but if you'd like to take a pass at it while you're in there it is essentially this:

When importing symbols from other files, the import places symbols onto pages with the same name in the slave file. So if my master file has 3 symbols on a page named `Symbols` and 10 symbols on a page called `Global UI`, it will import the 10 symbols on the Global UI page to a page with the same name in the slave file instead of importing all 13 symbols to the symbols page. 

The benefit to this is that it allows a work flow where I can have a few Master Symbol files for things like Global UI and Logos and such. The symbols on those pages in each doc will not be the main symbol page of my document. Which means when I create a new symbol in a slave file, it won't get cluttered with all of the global symbol assets I have for my project. In addition, if someone wanted everything on a singles Symbols page, they'd just leave their Symbols page named Symbols in each of their master symbol files and it would still work as the plugin does today.

I hope that makes sense haha.

In short, the way to handle this is when you actually add the symbol to the slave document, you grab the parentPage().name() for the source symbol, setCurrentPage() to that page in the slave document (or create if necessary) and then place the symbol on the page.

Anyway, that's how I handled it anyway. Not sure if you have a better idea of handling it but I of course can make this tweak myself after you finish this pull request.

**Comment by drewloomer** [comment]
@yourfavorite That sounds like an interesting behavior. I'd love to see the code you wrote to make this happen on the previous branch. I'd be happy to incorporate it.

I was able to get the import to fail by using your files (although the error was slightly different), and I've pushed a new fix. Can you check it out?

**Comment by drewloomer** [comment]
@yourfavorite I've implemented the parentPage stuff as per your suggestion. It seems to work well. Great idea! Hopefully the branch in its current state will work for you. Let me know what you find.

**Comment by yourfavorite** [comment]
Still getting errors. I am filtering in console.app to get these errors and for some reason it truncates the error and I don't know how to view the full thing. The first error I got is:
```
TypeError: instances.count is not a function. (In 'instances.count()', 'instances.count' is undefined)
line: 194
sourceURL: /Users/cbritt/Library/Application Support/com.bohemiancoding.sketch3/Plugins/Sketch Import Symbols/ImportSymbols.sketchplugin/Contents/Sketch/importSymbols.js
column: 27
stack: linkSymbols@/Users/cbritt/Library/Application Support/com.bohemiancoding.sketch3/Plugins/Sketch Import Symbols/ImportSymbols.sketchplugin/Contents/Sketch/importSymbols.js:194:27
addSymbol@/Users/cbritt/Library/Application Support/com.bohemiancoding.sketch3/Plugins/Sketch Import Symbols/ImportSymbols.sketchplugin/Contents/Sketch/importSymbols.js:173:14
addSymbols@/Users/cbritt/Library/Application Support/com.bohemiancoding.sketch3/Plugins/Sketch Import Symbols/ImportSymbols.sketchplugin/Contents/Sketch/importSymbols.js:139:18
startImport@/Users/cbritt/Library/Application Support/com.bohemiancoding.sketch3/Plugins/Sketch Import Symbols/ImportSymbols.sketchplugin/Contents/Sketch/importSymbols.js:122:25
impor<…>
```

I remember getting this error when I was doing .concat on MutableArray. So I changed line 106 to use arr.addObjectsFromArray() which fixed that problem. Then I get the following error which I haven't had a chance to dig into:
```
TypeError: undefined is not an object (evaluating 'arr.push')
line: 76
sourceURL: /Users/cbritt/Library/Application Support/com.bohemiancoding.sketch3/Plugins/Sketch Import Symbols/ImportSymbols.sketchplugin/Contents/Sketch/importSymbols.js
column: 7
stack: getNestedSymbolInstances@/Users/cbritt/Library/Application Support/com.bohemiancoding.sketch3/Plugins/Sketch Import Symbols/ImportSymbols.sketchplugin/Contents/Sketch/importSymbols.js:76:7
addSymbol@/Users/cbritt/Library/Application Support/com.bohemiancoding.sketch3/Plugins/Sketch Import Symbols/ImportSymbols.sketchplugin/Contents/Sketch/importSymbols.js:173:39
addSymbols@/Users/cbritt/Library/Application Support/com.bohemiancoding.sketch3/Plugins/Sketch Import Symbols/ImportSymbols.sketchplugin/Contents/Sketch/importSymbols.js:139:18
startImport@/Users/cbritt/Library/Application Support/com.bohemiancoding.sketch3/Plugins/Sketch Import Symbols/ImportSymbols.sketchplugin/Contents/Sketch/importSymbols.js:122:25
importSymbolsByID@/Users/cbritt/Libra<…>
```

**Comment by yourfavorite** [comment]
Oh man! I pulled the last batch of code from this pull request and not your new branch. MY BAD! Testing again :D

**Comment by yourfavorite** [comment]
Okay so the new code properly imports symbols from the source file on the correct page! However, there seems to be a bug. Let me know if you're seeing it too. The Symbols pages look great, but if you got Page 1 and look at the Nav Symbol on that page, it has not updated to the new symbol and shows the old symbol layout. And if you click the symbol override dropdown it shows two nav symbols even though there is only one on the symbols page. And if you double click on the nav on the page it takes you to the new nav updated with the new layout and colors... very strange.

Edit - So it looks like it added the new symbols to the correct pages but it left the old symbols and placed the new symbols directly on top of them so I didn't realize the old symbols were still there.

**Comment by drewloomer** [comment]
@yourfavorite Okay, it looks like "replaced" symbols weren't getting cleaned up and it was doing some WEIRD stuff that was jacking up the file. If you've saved `Test slave.sketch` after doing one of these bad imports, the file is probably pretty borked. But my most recent push to the new branch seems to work well, and symbols are being updated properly both on the symbols page and on Page 1 and Page 2. Let me know if it fixes your issues.

**Comment by yourfavorite** [comment]
Yeah I was a little worried about the old setValue on layer stuff messing things up. Looks like you're right. I did a clean import from the master file, then messed around with the symbols a bunch in both files. Reimported from master file and everything updated to match the master file. Both positioning, folders and symbol links are working great! Thanks so much for all the work on this Drew!

I believe @drewloomer's branch is good to merge into the master branch for this plugin @kmerc. Though I guess Drew needs to update this pull request.

**Comment by drewloomer** [comment]
Thanks for testing, @yourfavorite! I've updated the pull request. @kmerc let me know if you have any questions.

**Comment by drewloomer** [comment]
@kmerc in thinking more about this repo, would you be interested in moving it out of your personal account and into a shared repo so that you didn't have to be the sole gatekeeper? No biggie if not, just worried all these requests are getting annoying ;).

**Comment by yourfavorite** [comment]
@kmerc Sorry for the tag, just wanted to see if you had any thoughts on @drewloomer's last message. I'd like to start using this version of the plugin in some production files and would prefer the update has a home here before moving forward.

**Comment by drewloomer** [comment]
Seems like @kmerc is busy, so I've created a new package, repo and team for an updated symbol importer. https://github.com/sketch-symbol-importer/sketch-symbol-importer

Hope this doesn't ruffle anyones feathers, but I wanted to get these changes out to the world in an official capacity. Closing this PR.
